ansible_base/lib/utils/models.py

```diff
@@ -27,6 +27,19 @@ def get_type_for_model(model):
     return underscore(opts.object_name)
 
 
+def is_add_perm(codename: str) -> bool:
+    """Utility method to say whether codename represents adding permission.
+
+    This type of permission requires special treatment in several places
+    in DAB RBAC because the permission defers to the parent object.
+    Although the function is trivial, this standardizes the criteria.
+    """
+    use_codename = codename
+    if codename.count('.') == 1:
```

Comment:
:moyai: ?  What's with the dot?

Reply:
The reason is to cope with the Django format of permission. Here you can see DRF format permissions according to syntax like `'%(app_label)s.change_%(model_name)s'`, but it wasn't their invention. This syntax ultimately comes from Django. Clarifying - that is ultimately `"{app_label}.{codename}"`. As per other conversations here, this library won't support duplicate codenames, which are actually expected to be quite common (like `change_user` because of using a custom user model). While duplicate codenames are expected to _exist_ in auth.Permission, if you register two models with the same permission codename... it's almost certain that you did something wrong. Like, registering 2 user models, that should raise a red flag.

https://github.com/encode/django-rest-framework/blob/09a0c551ca934e486d2765e042a287c15fa85e75/rest_framework/permissions.py#L280-L289